### DISCUSSION TITLE
Reduce the requirements to verify a signed binary

### DISCUSSION THREAD

**Discussion by alalazo** [spack/spack#23656] [Ideas]
# Overview of the issue

Since binary repositories will presumably increase in importance in the Spack ecosystem, it becomes essential that verifying signatures poses the least requirements possible to avoid complicating the bootstrapping procedure.

Binaries are currently signed and verified using PGP (more specifically `gnupg`) but this software is quite complex and has many dependencies. To reduce the requirements on the side of verification the following strategies are being investigated:

1. Verify a PGP signature with a different tool (e.g. OpenSSL)
2. Permit to use tools different from PGP to sign and verify binaries in the hope that there are some as secure as PGP but simpler to bootstrap
3. Investigate ways to tune a build of `gnugp` that reduces as much as possible dependencies and requirements, but still works for Spack

## Facts to be considered

- Python implementations on almost all distros come with the standard `ssl` module, which is an interface to some SSL/TLS native library (OpenSSL on linux)
- This doesn't mean that on all systems where `python` is available the `openssl` or similar command line tools are available
- The reason is that the Python distribution only needs to be linked with `libssl` and `libcrypto`, and those libraries are usually packaged separately from both the command line tools and the development headers
- If `openssl` is available as a command line tool it should be possible to directly create a key pair in PEM format to sign and verify a file
- Using the `ssl` standard Python module may not be impossible, but requires to go  through an X.509 certificate while we are interested only in the associated private/public key pair

## Tools that can be used to sign and verify binaries

The following shows how to create a key pair, sign and verify a file with different tools. It also shows the complexity of the spec as concretized by Spack.

### GnuPG

[GnuPG](https://gnupg.org/) is the tool we currently use for signing and verifying binary caches. The downside of it is that it has a lot of dependencies, so it's difficult to bootstrap if not already present in the system:
```console
[    ]  gnupg@2.2.25%gcc@11.1.0 arch=linux-ubuntu18.04-broadwell
[bl  ]      ^libassuan@2.5.3%gcc@11.1.0 arch=linux-ubuntu18.04-broadwell
[bl  ]          ^libgpg-error@1.37%gcc@11.1.0 arch=linux-ubuntu18.04-broadwell
[b   ]              ^gawk@5.1.0%gcc@11.1.0 arch=linux-ubuntu18.04-broadwell
[bl  ]                  ^gettext@0.21%gcc@11.1.0+bzip2+curses+git~libunistring+libxml2+tar+xz arch=linux-ubuntu18.04-broadwell
[bl  ]                      ^bzip2@1.0.8%gcc@11.1.0~debug~pic+shared arch=linux-ubuntu18.04-broadwell
[b   ]                          ^diffutils@3.7%gcc@11.1.0 arch=linux-ubuntu18.04-broadwell
[bl  ]                              ^libiconv@1.16%gcc@11.1.0 arch=linux-ubuntu18.04-broadwell
[bl  ]                      ^libxml2@2.9.10%gcc@11.1.0~python arch=linux-ubuntu18.04-broadwell
[b   ]                          ^pkgconf@1.7.4%gcc@11.1.0 arch=linux-ubuntu18.04-broadwell
[blr ]                          ^xz@5.2.5%gcc@11.1.0~pic libs=shared,static arch=linux-ubuntu18.04-broadwell
[bl  ]                          ^zlib@1.2.11%gcc@11.1.0+optimize+pic+shared arch=linux-ubuntu18.04-broadwell
[bl  ]                      ^ncurses@6.2%gcc@11.1.0~symlinks+termlib abi=none arch=linux-ubuntu18.04-broadwell
[bl  ]                      ^tar@1.34%gcc@11.1.0 arch=linux-ubuntu18.04-broadwell
[bl  ]                  ^gmp@6.2.1%gcc@11.1.0 arch=linux-ubuntu18.04-broadwell
[b   ]                      ^autoconf@2.69%gcc@11.1.0 arch=linux-ubuntu18.04-broadwell
[b r ]                          ^m4@1.4.18%gcc@11.1.0+sigsegv patches=3877ab548f88597ab2327a2230ee048d2d07ace1062efe81fc92e91b7f39cd00,fc9b61654a3ba1a8d6cd78ce087e7c96366c290bc8d2c299f09828d793b853c8 arch=linux-ubuntu18.04-broadwell
[bl  ]                              ^libsigsegv@2.13%gcc@11.1.0 arch=linux-ubuntu18.04-broadwell
[blr ]                          ^perl@5.32.1%gcc@11.1.0+cpanm+shared+threads arch=linux-ubuntu18.04-broadwell
[bl  ]                              ^berkeley-db@18.1.40%gcc@11.1.0+cxx~docs+stl patches=b231fcc4d5cff05e5c3a4814f6a5af0e9a966428dc2176540d2c05aff41de522 arch=linux-ubuntu18.04-broadwell
[bl  ]                              ^gdbm@1.19%gcc@11.1.0 arch=linux-ubuntu18.04-broadwell
[bl  ]                                  ^readline@8.1%gcc@11.1.0 arch=linux-ubuntu18.04-broadwell
[b   ]                      ^automake@1.16.3%gcc@11.1.0 arch=linux-ubuntu18.04-broadwell
[b   ]                      ^libtool@2.4.6%gcc@11.1.0 arch=linux-ubuntu18.04-broadwell
[bl  ]                  ^mpfr@4.1.0%gcc@11.1.0 arch=linux-ubuntu18.04-broadwell
[b   ]                      ^autoconf-archive@2019.01.06%gcc@11.1.0 arch=linux-ubuntu18.04-broadwell
[b   ]                      ^texinfo@6.5%gcc@11.1.0 patches=12f6edb0c6b270b8c8dba2ce17998c580db01182d871ee32b7b6e4129bd1d23a,1732115f651cff98989cb0215d8f64da5e0f7911ebf0c13b064920f088f2ffe1 arch=linux-ubuntu18.04-broadwell
[bl  ]      ^libgcrypt@1.9.1%gcc@11.1.0 arch=linux-ubuntu18.04-broadwell
[bl  ]      ^libksba@1.4.0%gcc@11.1.0 arch=linux-ubuntu18.04-broadwell
[bl  ]      ^npth@1.6%gcc@11.1.0 arch=linux-ubuntu18.04-broadwell
[  r ]      ^pinentry@1.1.0%gcc@11.1.0 arch=linux-ubuntu18.04-broadwell

```
This is an example that shows how to generate a key, sign and verify a file:
```console
$ # Generate a key pair (RSA-4096)
$ gpg --full-gen-key
gpg (GnuPG) 2.2.4; Copyright (C) 2017 Free Software Foundation, Inc.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Please select what kind of key you want:
   (1) RSA and RSA (default)
   (2) DSA and Elgamal
   (3) DSA (sign only)
   (4) RSA (sign only)
Your selection? 
[ ... ]

$ # Sign a tarball with gpg
$ gpg --detach-sign --armor --default-key A4EDAB45 --output opengpg/linux-rhel5-x86_64-gcc-9.3.0-clingo-bootstrap-spack-ffoyoewfd6pdwbjniodfkqusyvkrbhyi.tar.gz.asc linux-rhel5-x86_64-gcc-9.3.0-clingo-bootstrap-spack-ffoyoewfd6pdwbjniodfkqusyvkrbhyi.tar.gz

$ # Verify a signature
$ gpg --verify opengpg/linux-rhel5-x86_64-gcc-9.3.0-clingo-bootstrap-spack-ffoyoewfd6pdwbjniodfkqusyvkrbhyi.tar.gz.asc linux-rhel5-x86_64-gcc-9.3.0-clingo-bootstrap-spack-ffoyoewfd6pdwbjniodfkqusyvkrbhyi.tar.gz 
gpg: Signature made ven 14 mag 2021 18:55:10 CEST
gpg:                using RSA key DBF789A6FFA961FC6AC6E920F198CA81A4EDAB45
gpg: Good signature from "Massimiliano Culpo (Example of signature/verification) <massimiliano.culpo@gmail.com>" [ultimate]

$ # Export the public key for others to import
$ gpg --export --armor --output opengpg/public_key.pgp A4EDAB45

```
### Minisign

[Minisign](https://jedisct1.github.io/minisign/) is a simple tool to sign files and verify signatures. It can complement OpenPGP in Spack to sign binaries. The package:
1. Depends only on libsodium and the executable can be statically linked and distributed alone
2. Encryption is based on Ed25519, which is slightly less secure than RSA-4096, but reported to be much faster

```console
[    ]  minisign@0.9%gcc@11.1.0~ipo build_type=RelWithDebInfo arch=linux-ubuntu18.04-broadwell
[b   ]      ^cmake@3.20.2%gcc@11.1.0~doc+ncurses+openssl+ownlibs~qt build_type=Release arch=linux-ubuntu18.04-broadwell
[bl  ]      ^libsodium@1.0.18%gcc@11.1.0 arch=linux-ubuntu18.04-broadwell
```
The dependency on libsodium can be made a pure build dependency if libsodium is compiled statically. This would be fine from the modeling perspective since `minisign` just provides a command line tools and no other libraries that could be reused in upstream projects. Here are a [few considerations](https://security.stackexchange.com/questions/143083/ssh-key-strength-factor-besides-key-length-say-ed25519-vs-rsa-4096) on the security of the Ed25519 algorithm. 

This is an example that shows how to generate a key, sign and verify a file:
```console
$ # Generate a key pair (Ed25519)
$ minisign -G
Please enter a password to protect the secret key.

Password: 
Password (one more time): 
Deriving a key from the password in order to encrypt the secret key... done

The secret key was saved as /home/culpo/.minisign/minisign.key - Keep it secret!
The public key was saved as minisign.pub - That one can be public.

Files signed using this key pair can be verified with the following command:

minisign -Vm <file> -P RWQSv4CMHQj1fnqFcDjv5Vkzot8EE37cL/cZNTK7tYyJ5I5Qng8JPjUy

$ # Sign a tarball with minisign
$ minisign -S -x minisign/linux-rhel5-x86_64-gcc-9.3.0-clingo-bootstrap-spack-ffoyoewfd6pdwbjniodfkqusyvkrbhyi.tar.gz.minisig -m linux-rhel5-x86_64-gcc-9.3.0-clingo-bootstrap-spack-ffoyoewfd6pdwbjniodfkqusyvkrbhyi.tar.gz 
Password: 
Deriving a key from the password and decrypting the secret key... done

$ # Verify a signature
$ minisign -V -m linux-rhel5-x86_64-gcc-9.3.0-clingo-bootstrap-spack-ffoyoewfd6pdwbjniodfkqusyvkrbhyi.tar.gz -x minisign/linux-rhel5-x86_64-gcc-9.3.0-clingo-bootstrap-spack-ffoyoewfd6pdwbjniodfkqusyvkrbhyi.tar.gz.minisig 
Signature and comment signature verified
Trusted comment: timestamp:1621012335	file:linux-rhel5-x86_64-gcc-9.3.0-clingo-bootstrap-spack-ffoyoewfd6pdwbjniodfkqusyvkrbhyi.tar.gz

$ # Export the public key for others to import
$ minisign -R -s ~/.minisign/minisign.key -p minisign/minisign.pub
```

### OpenSSL

[OpenSSL](https://www.openssl.org/) is a tool that is broadly available on clusters:
```console
[    ]  openssl@1.1.1k%gcc@11.1.0~docs+systemcerts arch=linux-ubuntu18.04-broadwell
[b   ]      ^perl@5.32.1%gcc@11.1.0+cpanm+shared+threads arch=linux-ubuntu18.04-broadwell
[bl  ]          ^berkeley-db@18.1.40%gcc@11.1.0+cxx~docs+stl patches=b231fcc4d5cff05e5c3a4814f6a5af0e9a966428dc2176540d2c05aff41de522 arch=linux-ubuntu18.04-broadwell
[bl  ]          ^gdbm@1.19%gcc@11.1.0 arch=linux-ubuntu18.04-broadwell
[bl  ]              ^readline@8.1%gcc@11.1.0 arch=linux-ubuntu18.04-broadwell
[bl  ]                  ^ncurses@6.2%gcc@11.1.0~symlinks+termlib abi=none arch=linux-ubuntu18.04-broadwell
[b   ]                      ^pkgconf@1.7.4%gcc@11.1.0 arch=linux-ubuntu18.04-broadwell
[bl  ]      ^zlib@1.2.11%gcc@11.1.0+optimize+pic+shared arch=linux-ubuntu18.04-broadwell
```
Its cryptography layer can be used to sign and verify files:
```console
$ # Generate a key pair (RSA 4096)
$ openssl genrsa -aes128  -out openssl/private.pem 4096
Generating RSA private key, 4096 bit long modulus (2 primes)
..........................................................................................................................................................................................................................................................................................++++
...++++
e is 65537 (0x010001)
Enter pass phrase for openssl/private.pem:
Verifying - Enter pass phrase for openssl/private.pem:

$ # Sign a tarball with OpenSSL
$ openssl dgst -sha256 -sign openssl/private.pem -out openssl/sign.sha256 linux-rhel5-x86_64-gcc-9.3.0-clingo-bootstrap-spack-ffoyoewfd6pdwbjniodfkqusyvkrbhyi.tar.gz 
Enter pass phrase for openssl/private.pem:

$ # Export the public key for others to import
$ openssl rsa -in openssl/private.pem -pubout -out openssl/public.pem
Enter pass phrase for openssl/private.pem:
writing RSA key

$ # Verify a signature
$ openssl dgst -sha256 -verify openssl/public.pem -signature openssl/sign.sha256 linux-rhel5-x86_64-gcc-9.3.0-clingo-bootstrap-spack-ffoyoewfd6pdwbjniodfkqusyvkrbhyi.tar.gz 
Verified OK
```

## Converting keys

There are a few articles that explain how to convert key pairs to different formats. An interesting project is [monkeysphere](https://web.monkeysphere.info/) which provides commandline tools like `openpgp2pem` or `openpgp2ssh`. It may be complicated though to match the algorithms among tools. 

## Python modules to sign/verify files

There are a few Python modules to help signing and verifying files, but they all have dependencies on [cryptography](https://cryptography.io/en/latest/) which in turn imposes a dependency on clients to have a Rust compiler installed. This defeats the purpose of simplyfying the bootstrapping method for `pgp`.

A pure python implementation of some ECDSA algorithms can be found in the [ecdsa](https://github.com/tlsfuzzer/python-ecdsa) package but _it is NOT meant to be a secure implementation_ and they point users back to `cryptography` in case a secure implementation is needed.

**Comment by haampie** [comment]
The huge list of gpg packages is a bit misleading since gawk is pulling in most of them, which is just a build dep. Maybe another consideration is whether it even builds on windows?

**Comment by haampie** [comment]
Let me point out some more issues I see with GPG:

- It likes to store the private key on disk as you have to run `gpg --import`:
  
  This may be a security vulnerability when signing builds in automated pipelines on shared systems or during docker build, where you don't want to leave traces on the filesystem. It would be great if the key could be stored as an env variable, read from stdin, and never be persisted to the filesystem.
- It has runtime config files: are we sure system config is not pulled in influencing code signing?
- gnupg2 uses gpg-agent, which is a daemon that may perform password caching in case you password-protect your private keys (which you should on your own system!); this may be something users may not want to use, but is currently enforced -- see [the discussion here](https://stackoverflow.com/questions/47273922/purpose-of-gpg-agent-in-gpg2). Only gnupg1 can disable all helper programs, but does not support modern elliptic curve cryptography.
- This is more my personal gripe, but gnupg2 is not binary relocatable as it hardcodes absolute paths to helper programs such as gpg-agent without overrides through environment variables (let's say spack's definition of relocatability is a weak one :p I'm talking about binary relocation with relative rpaths where `mv` should not break binaries). Only gnupg1 is binary relocatable.
